src/Controller/FavoritesController.php

```diff
@@ -0,0 +1,189 @@
+      ],
+      'taxonomy_term' => [
+        'gc_category' => [],
+      ],
+    ];
+
+    $uid = $this->currentUser()->id();
+    $query = $this->database->select('vy_favorite_item', 'f');
+    $query->condition('f.uid', $uid);
+    $query->fields('f', [
+      'id',
+      'ref_entity_type',
+      'ref_entity_bundle',
+      'ref_entity_id',
+    ]);
+    $result = $query->execute()->fetchAll();
+    if (empty($result)) {
+      return new JsonResponse($data);
+    }
+
+    foreach ($result as $item) {
+      $data[$item->ref_entity_type][$item->ref_entity_bundle][] = [
+        'id' => $item->id,
+        'entity_id' => $item->ref_entity_id,
+      ];
+    }
+
+    return new JsonResponse($data);
+  }
+
+  /**
+   * Add item to Favorites list.
+   *
+   * @param \Symfony\Component\HttpFoundation\Request $request
+   *   The request.
+   *
+   * @return \Symfony\Component\HttpFoundation\JsonResponse
+   *   Json Array with favorites items grouped by entity type and bundle.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+   * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+   * @throws \Drupal\Core\Entity\EntityStorageException
+   */
+  public function add(Request $request) {
+    $data = $this->getData($request);
+    if ($data instanceof JsonResponse) {
+      return $data;
+    }
+
+    $favorite_item = $this->entityTypeManager()
+      ->getStorage('vy_favorite_item')
+      ->create([
+        'ref_entity_type' => $data['entity_type'],
+        'ref_entity_bundle' => $data['entity_bundle'],
+        'ref_entity_id' => $data['entity_id'],
+      ]);
+    $favorite_item->save();
+    // Return full list of favorites.
+    return $this->list();
+  }
+
+  /**
+   * Remove item from the Favorites list.
+   *
+   * @param \Symfony\Component\HttpFoundation\Request $request
+   *   The request.
+   *
+   * @return \Symfony\Component\HttpFoundation\JsonResponse
+   *   Json Array with favorites items grouped by entity type and bundle.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+   * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+   * @throws \Drupal\Core\Entity\EntityStorageException
+   */
+  public function delete(Request $request) {
+    $data = $this->getData($request);
+    if ($data instanceof JsonResponse) {
+      return $data;
+    }
+    $favorite_items = $this->entityTypeManager()
+      ->getStorage('vy_favorite_item')
+      ->loadByProperties([
+        'uid' => $this->currentUser()->id(),
+        'ref_entity_type' => $data['entity_type'],
+        'ref_entity_bundle' => $data['entity_bundle'],
+        'ref_entity_id' => $data['entity_id'],
+      ]);
+    if (empty($favorite_items)) {
+      return new JsonResponse(['message' => 'Item not exists']);
+    }
+    foreach ($favorite_items as $item) {
+      $item->delete();
+    }
+
+    // Return full list of favorites.
+    return $this->list();
```

Comment:
same here

Reply:
Fixed in https://github.com/fivejars/openy_gated_content/pull/231